benchmark/benchmark.hpp

```diff
@@ -441,20 +447,141 @@ class multiple_malloc_free_benchmark : public benchmark_interface<Size, Alloc> {
+                          free.back());
+                auto offset = free.back();
+                free.pop_back();
+
+                n.push_back({true, offset, sizeGenerator.nextSize()});
+                allocated.push_back(offset);
+            } else {
+                std::swap(allocated[dist_offset(generator) % allocated.size()],
+                          allocated.back());
+                auto offset = allocated.back();
+                allocated.pop_back();
+
+                n.push_back({false, offset, 0});
+                free.push_back(offset);
+            }
         }
+
         next_iter[tid] = std::make_unique<next_alloc_data_iterator>(n.cbegin());
     }
 };
+
+// This class benchmarks performance randomly allocates and frees,
+// Firstly slowly increasing memory footprint, and later decreasing
+template <
+    typename Size, typename Alloc,
+    typename =
+        std::enable_if_t<std::is_base_of<alloc_size_interface, Size>::value>,
+    typename =
+        std::enable_if_t<std::is_base_of<allocator_interface, Alloc>::value>>
+class peak_alloc_benchmark
+    : public multiple_malloc_free_benchmark<Size, Alloc> {
+    using base = multiple_malloc_free_benchmark<Size, Alloc>;
+    virtual void prepareWorkload(benchmark::State &state) override {
+
+        auto tid = state.thread_index();
+        auto &n = this->next[tid];
+        std::default_random_engine generator;
+        std::uniform_int_distribution<size_t> dist_offset(0,
+                                                          this->max_allocs - 1);
+        std::uniform_real_distribution<double> dist_opt_type(0, 1);
+        generator.seed(0);
+        auto sizeGenerator = this->alloc_sizes[tid];
+
+        n.clear();
+        std::vector<size_t> free;
+        std::vector<size_t> allocated;
+        // we start without any allocations
+        for (size_t i = 0; i < this->max_allocs; i++) {
+            free.push_back(i);
+        }
+
+        int64_t iterations = state.max_iterations * this->allocsPerIterations;
+        for (int64_t j = 0; j < iterations; j++) {
+            int64_t target_allocation;
+            int64_t max_allocs = static_cast<int64_t>(this->max_allocs);
+            if (j < iterations / 2) {
+                target_allocation = 2 * max_allocs * j / iterations;
+            } else {
+                target_allocation =
+                    -2 * max_allocs * j / iterations + 2 * max_allocs;
+            }
+
+            auto x = static_cast<double>(target_allocation -
+                                         static_cast<double>(allocated.size()));
+            // high sigma value cause small changes in probability between alloc and free
+            // based on offset x from target_allocation number.
+            // sigma == 1000 causes that actual number of allocation are +/-40 of target number
+            const double sigma = 1000;
+            auto cdf = normalCDF(x, sigma);
```

Comment:
what is CDF?

Reply:
cumulative distribution function, or "dystrybuanta" in polish